File: go/internal/sol/p0017/p0017.go
package p0017

import (
	"strconv"
)

var under20 = []int{
	len(""),
	len("one"),
	len("two"),
	len("three"),
	len("four"),
	len("five"),
	len("six"),
	len("seven"),
	len("eight"),
	len("nine"),
	len("ten"),
	len("eleven"),
	len("twelve"),
	len("thirteen"),
	len("fourteen"),
	len("fifteen"),
	len("sixteen"),
	len("seventeen"),
	len("eighteen"),
	len("nineteen"),
}

var mults10 = []int{
	len(""),
	len(""),
	len("twenty"),
	len("thirty"),
	len("forty"),
	len("fifty"),
	len("sixty"),
	len("seventy"),
	len("eighty"),
	len("ninety"),
}

func countLetters(n int) int {
	if n <= 0 || n > 1000 {
		panic("range error")
	}

	switch {
	case n == 1000:
		return under20[1] + len("thousand")
	case n < 20:
		return under20[n]
	case n < 100:
		return mults10[n/10] + under20[n%10]
	default:
		q, r := n/100, n%100
		if r == 0 {
			return under20[q] + len("hundred")
		} else {
			return under20[q] + len("hundred") + len("and") + countLetters(r)
		}
	}
}

func compute(limit int) string {
	var result int
	for i := 1; i <= limit; i++ {
		result += countLetters(i)
	}

	return strconv.FormatInt(int64(result), 10)
}

func Solve() string {
	return compute(1_000)
}

File: haskell/src/Sol/P0017.hs
module Sol.P0017 (compute, solve) where

import Data.Array.Unboxed (UArray, listArray, (!))

under20 :: UArray Int Int
under20 =
    listArray (0, 19) $
        length
            <$> [ ""
                , "one"
                , "two"
                , "three"
                , "four"
                , "five"
                , "six"
                , "seven"
                , "eight"
                , "nine"
                , "ten"
                , "eleven"
                , "twelve"
                , "thirteen"
                , "fourteen"
                , "fifteen"
                , "sixteen"
                , "seventeen"
                , "eighteen"
                , "nineteen"
                ]

mults10 :: UArray Int Int
mults10 =
    listArray (0, 9) $
        length
            <$> [ ""
                , ""
                , "twenty"
                , "thirty"
                , "forty"
                , "fifty"
                , "sixty"
                , "seventy"
                , "eighty"
                , "ninety"
                ]

countLetters :: Int -> Int
countLetters n
    | n <= 0 || n > 1000 = error "range error"
    | n == 1000 = (under20 ! 1) + (length "thousand")
    | n < 20 = under20 ! n
    | n < 100 = (mults10 ! (div n 10)) + (under20 ! (mod n 10))
    | otherwise =
        let (q, r) = (div n 100, mod n 100)
         in case r of
                0 -> (under20 ! q) + (length "hundred")
                _ -> (under20 ! q) + (length "hundred") + (length "and") + (countLetters r)

compute :: Int -> String
compute limit =
    show . sum $ countLetters <$> [1 .. limit]

solve :: String
solve = compute 1_000
